proptools/compare.go

```diff
@@ -0,0 +1,248 @@
+func compareTypes(a namedValue, b namedValue) (comparison Comparison) {
+	aType, aTypeText := getType(a.item)
+	bType, bTypeText := getType(b.item)
+	if aType == bType {
+		return equal
+	}
+	return different(a.name + " is of type " + aTypeText + " whereas " + b.name + " is of type " + bTypeText)
+}
+
+// primary method that compares whether two arbitrary values are equal
+func deepCompareGeneric(a namedValue, b namedValue, depth int) (comparison Comparison) {
+	depth++
+	if depth > 50 {
+		panic(fmt.Sprintf("Detected probable infinite loop comparing %s (%s) and %s (%s). This diff checker doesn't yet support data structures having cycles.",
+			a.name,
+			a.item,
+			b.name,
+			b.item))
+	}
+
+	typeComparison := compareTypes(a, b)
+	if !typeComparison.Equal {
+		return typeComparison
+	}
+
+	switch a.item.Kind() {
+	case reflect.Array, reflect.Slice:
+		return deepCompareArrays(a, b, depth)
+	case reflect.Map:
+		return deepCompareMaps(a, b, depth)
+	case reflect.Struct:
+		return deepCompareStructs(a, b, depth)
+	case reflect.Ptr, reflect.Interface:
+		return deepComparePointers(a, b, depth)
+	case reflect.Int, reflect.String, reflect.Bool:
+		return comparePrimitives(a, b, depth)
+	default:
+		panic(fmt.Sprintf("unrecognized type %v of value %#v\n", a.item.Kind(), a.item))
+
+	}
+}
+
+// compares two pointers in the same manner as reflect.deepEqual
+func deepComparePointers(a namedValue, b namedValue, depth int) (comparison Comparison) {
+	if a.item.IsNil() || b.item.IsNil() {
+		if a.item.IsNil() && b.item.IsNil() {
+			return equal
+		} else {
+
+			if a.item.IsNil() {
+				return different(fmt.Sprintf("%s is nil whereas %s = %#v", a.name, b.name, b.item))
+			} else {
+				return different(fmt.Sprintf("%s is nil whereas %s = %#v", b.name, a.name, a.item))
+			}
+		}
+
+	}
+	return deepCompareGeneric(namedValue{a.name, a.item.Elem()},
+		namedValue{b.name, b.item.Elem()},
+		depth)
+}
+
+// compares two non-struct items
+func comparePrimitives(actual namedValue, expected namedValue, depth int) (comparison Comparison) {
+	// TODO can we find a simpler implementation of this method that still works with private fields?
+	var a, b interface{}
+	if actual.item.Kind() != expected.item.Kind() {
+		return different(fmt.Sprintf("%s is of type %s whereas %s is of type %s", actual.name, actual.item.Kind(), expected.name, expected.item.Kind()))
+	}
+	switch actual.item.Kind() {
+	case reflect.String:
+		a = actual.item.String()
+		b = expected.item.String()
+	case reflect.Bool:
+		a = actual.item.Bool()
+		b = expected.item.Bool()
+	case reflect.Int:
+		a = actual.item.Int()
+		b = expected.item.Int()
+	default:
+		panic(fmt.Sprintf("unrecognized types, %s (%#v) and %s (%#v)", actual.name, actual.item, expected.name, expected.item))
+	}
+	if reflect.DeepEqual(a, b) {
+		return equal
+	} else {
+		return different(fmt.Sprintf("%s = %#v whereas %s = %#v", actual.name, actual.item, expected.name, expected.item))
+	}
+}
+
+// compares two maps
```

Comment:
document the semantics of this comparison, especially as compared to reflect.DeepEqual